multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitializationFilter.java

```diff
@@ -0,0 +1,172 @@
+    this.keycloakProfileRetriever = keycloakProfileRetriever;
+    this.tokenExtractor = tokenExtractor;
+    this.permissionChecker = permissionChecker;
+    this.keycloakSettings = settings;
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) {}
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
+      throws IOException, ServletException {
+    try {
+      super.doFilter(request, response, filterChain);
+    } catch (MachineTokenJwtException mte) {
+      filterChain.doFilter(request, response);
+    } catch (JwtException e) {
+      sendError(response, SC_UNAUTHORIZED, e.getMessage());
+      return;
+    }
+  }
+
+  @Override
+  protected String getUserId(String token) {
+    Claims claims = jwtParser.parseClaimsJws(token).getBody();
+    return claims.getSubject();
+  }
+
+  @Override
+  public Subject extractSubject(String token) throws ServletException {
+
+    Jws<Claims> jwt = jwtParser.parseClaimsJws(token);
+    Claims claims = jwt.getBody();
+    LOG.debug("JWT = {}", jwt);
+    // OK, we can trust this JWT
+
+    try {
+      String username =
+          claims.get(
+              keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING), String.class);
+      if (username == null) { // fallback to unique id promised by spec
+        // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
+        username = claims.getIssuer() + ":" + claims.getSubject();
+      }
+      String id = claims.getSubject();
+
+      String email =
+          retrieveEmail(token, claims, username, id)
+              .orElseThrow(
+                  () ->
+                      new JwtException(
+                          "Unable to authenticate user because email address is not set in keycloak profile"));
+      User user = userManager.getOrCreateUser(id, email, username);
+      return new AuthorizedSubject(
+          new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
+    } catch (ServerException | ConflictException e) {
+      throw new ServletException(
+          "Unable to identify user " + claims.getSubject() + " in Che database", e);
+    }
+  }
+
+  @Override
+  protected void handleMissingToken(
+      ServletRequest request, ServletResponse response, FilterChain chain) throws IOException {
+    sendError(response, 401, "Authorization token is missed");
+  }
+
+  private Optional<String> retrieveEmail(String token, Claims claims, String username, String id)
+      throws ServerException {
+    String email = claims.get("email", String.class);
+
+    if (isNullOrEmpty(email)) {
+      boolean userNotFound = false;
+      try {
+        userManager.getById(id);
+      } catch (NotFoundException e) {
+        userNotFound = true;
+      }
+      if (userNotFound) {
+        try {
+          EnvironmentContext.getCurrent().setSubject(new SubjectImpl(username, id, token, true));
```

Comment:
This seems quite oddly placed.. I would personally not expect setting the one most important global variable in a method called `retrieveEmail`.

Reply:
This is done to make correctly signed request to keycloak to obtain user profile. I will try to look at another possible solution.

Reply:
So i don't think i can do somehing completely different here, so i simplified code a bit and add an comment why it's require context to be set here, hope is ok @metlos ?

Reply:
What about 
```java
       KeycloakProfileRetriever keycloakProfileRetriever = new KeycloakProfileRetriever(keycloakSettings, new DefaultHttpJsonRequestFactory(){
            @Override
            public HttpJsonRequest fromUrl(@NotNull String url) {
              return super.fromUrl(url)
                      .setAuthorizationHeader(
                              "Bearer " + token);
            }

            @Override
            public HttpJsonRequest fromLink(@NotNull Link link) {
              return super.fromLink(link)
                      .setAuthorizationHeader(
                              "Bearer " +token);
            }
          });




          Map<String, String> profileAttributes =
              keycloakProfileRetriever.retrieveKeycloakAttributes();
```
or something like that.

Reply:
what about 
```java
public class KeycloakProfileRetriever {
....
  public Map<String, String> retrieveKeycloakAttributes(@NotNull String authorizationHeader)
      throws ServerException {
    try {
      return requestFactory
          .fromUrl(keyclockCurrentUserInfoUrl)
          .setAuthorizationHeader(authorizationHeader)
          .request()
          .asProperties();
    } catch (IOException | ApiException e) {
      LOG.warn("Exception during retrieval of the Keycloak user profile", e);
      throw new ServerException("Exception during retrieval of the Keycloak user profile", e);
    }
  }
```